rfcs/79-minimal-fragmenting-relationships.md

```diff
@@ -0,0 +1,126 @@
+
+Similarly systems cache their `ArchetypeComponentId` accesses for the purposes of multithreading, we also need to update theses caches to remove deleted ids presenting many of the same problems. 
+##### Access Bitsets and Component SparseSets
+The `FixedBitSet` implementation used by `Access` inside of systems, system params, queries etc. relies on dense component ids to avoid using a large amount of memory. If we were to start inserting ids with arbitrarily large entity targets as well as component ids shifted into the upper bits each `FixedBitSet` would allocate non-trivial amounts of memory to have flags for all the intervening ids. In order to minimize the affects of this we could consider all `(R, *)` pairs as one id in accesses, reducing the amount we need to track at the expense of some theoretical ability to parallelise.
+
+Similarly the `SparseSet` implementation used in both table and sparse component storage also operate under the assumption that component ids will remain low and would allocate an even larger amount of memory storing indexes for all the intervening ids when storing relationship pairs on an entity.
+
+The most practical way to solve this is using paged versions of the appropriate data structures, for bitsets this can take the form of existing hierarchical or compressed bitsets like `roaring`.
+### Minimal Implementation
+Implementing the above in order of dependance could look like:
+- Resolve [#12144](https://github.com/bevyengine/bevy/issues/12144) so entities can be persisted in the render world.
+- Merge observers [#10839](https://github.com/bevyengine/bevy/pull/10839)
+- Refactor queries to become entities with caches updated via observers
+- Implement archetype deletion and the associated clean-up for systems and queries 
+- Replace `FixedBitSet` and `SparseSet` implementations with ones that support sparse ids
+
+After all the issues above are addressed the actual implementation of the feature can begin. This requires changes in 3 main areas:
+- Implementation of `WorldData` and `WorldFilter` implementations for accessing relationship targets
+- Transition to new `Identifier` implemented in [#9797](https://github.com/bevyengine/bevy/pull/9797) instead of `ComponentId` in all places where pairs could be expected: tables, archetypes etc. 
+- New methods for `EntityCommands`, `EntityMut`/`Ref`/`WorldMut` and `QueryBuilder`
+## Drawbacks
+- Increased archetype fragmentation when in use
```

Comment:
Increasing archetype fragmentation has an impact on memory fragmentation and query iteration performance, both of which should be detailed here.

Reply:
The impact of fragmentation is complex:
- there is a wide range of different scenarios with wildly different performance characteristics
- there are many scenarios where fragmentation can improve query performance
- component fragmentation can be kept in check because of the table/archetype split
- there's prior art for a [non-fragmenting relationship implementation](https://www.flecs.dev/flecs/md_docs_2Relationships.html#union-property) which can coexist with a fragmenting one

The flecs vs. bevy hierarchy benchmarks (which highlights just one narrow aspect of relationships) show that the impact is not uniformly good or bad, with narrow trees being faster without fragmenting relationships, and wide trees being (much) faster:

<img width="556" alt="Screenshot 2024-04-03 at 9 43 42 AM" src="https://github.com/bevyengine/rfcs/assets/9919222/2c15b94f-2359-421f-815b-59600e9897ff">

(flecs left, bevy right, number is flecs / bevy)

tl;dr when calling this out it's easy to focus on the negatives of fragmentation which are easy to understand without balancing that out with the positives, which are harder to understand and measure without an implementation, but do exist and have been proven.